### ISSUE TITLE
Use `__main__.py` to inclose `_backup.py` etc..

### ISSUE THREAD

**Issue by larkost** [rethinkdb/rethinkdb#5147]
Our current way of running the administrative scripts such as `export`, `restore`, and `index_rebuild` is a bit awkward in that you call `export backup` and the server uses `execvp` to call `rethinkdb-export`. That is installed only when you have the driver installed on the system, so uses the first on on `$PATH`, and uses `pkg_resources.load_entry_point` to actually start the process from the module, again found via the `$PATH`. This all makes it hard to test, and a fragile if you are in the middle of transferring from one version of RethinkDB to another.

If instead we included a `__main__.py` file, and then drove all of the commands out of that one it would mean that the driver could be called directly, even if it was not on the path (e.g.: `python -m /path/to/driver`). That would change the call from the `rethinkdb` binary to be `python -m rethinkdb export` to match what we have now. And when combined with #4507 it would use the corresponding version of the driver.

**Comment by larkost** [comment]
Note: if we do get rid of the `entry_points`, we would have to make sure that they get cleaned up when someone updated the python driver.

**Comment by danielmewes** [comment]
Side note: Please don't rename issues unless the title has become fundamentally wrong. https://github.com/rethinkdb/company/issues/71

**Comment by danielmewes** [comment]
From what I can tell, this sounds like a good improvement to avoid one type of inconsistency when running the backup scripts. Basically it makes sure that the driver we use matches the backup script we run, correct?

**Comment by larkost** [comment]
Yes, it closes one (unlikely) hole. But more importantly it allows for specifying full paths when calling the "scripts" making testing and development a lot easier. It also opens the door to driving the scripts directly without calling the `rethinkdb` binary directly, which could mean you call backup from another computer (not that this would be a good idea from an i/o perspective).

**Comment by larkost** [comment]
I am working on this since it makes testing some changes to these functions a lot easier to test.

**Comment by larkost** [comment]
An initial round of this was merged into `next` with df57248c (and `v2.3.x` in 979600e7) after CR 3608 from branch `larkost/5147-python-driver-utils`. There is some additional work to do that I am going to continue do in this bug.

**Comment by larkost** [comment]
A second round has also been added to `next` with d9a31bc4 after CR 3643. It cleans up dump/import and has the side effect of making them slightly (up to 15% in my test case) faster. Most notably the semi-opaque `rdb_call_wrapper` function has been simplified into a per-query function that is a lot easier to read.

Two more things remain to be done before I call this one done:
1. An experiment to see if combining the reader and writer processes into a cluster of processes that do both yields any speedup.
2. Applying the same lessons learned to the `dump`/`export` routines. Note that a lot of the work of straightening these out has already made it in here.

**Comment by larkost** [comment]
The experiment for 1 has run its course. The idea was to combine the reader and writer into a single process, and then run multiple of them against a file simultaneously. While only one of them could read the file at a single time, hopefully the time spent processing and uploading the information could overlap with others. Sadly it looks like the reading of the file takes too much time at the moment (we have to process JSON in doing it), so we don't wind up with enough overlap to make this approach worth it.

On my (very simple) test set this approach only wound up with near-parity (a couple of percent slower) after a bit of tuning for the machines I was running on. Since this does not get near the bar of performing significantly faster I am pulling the plug on the experiment.

Since it is possible that I have messed up something killing my performance, or that something could change in this calculus in the future, I am saving the branch as `larkost/5147-worker-test`.

**Comment by larkost** [comment]
Work on merging the cleanup from the experiment for `_import` has been merged into `next` with 9e1a71c0 after review in CR 3670 (from branch larkost/5147-python-driver-utils-part3). Mostly this is organizational cleanup, but I did see a small speedup in my in-development testing.

So only the cleanup of `_export`/`_dump` is left. From my looks at that there is less to do, so it should be much faster.
